File: Python/ByRound/2053/2053_H_Delicate_Anti_monotonous_Operations.py
# Problem: CF 2053 H - Delicate Anti-monotonous Operations
# https://codeforces.com/contest/2053/problem/H

"""
Algorithm: Delicate Anti-monotonous Operations
Purpose: To find the maximum possible sum of array elements after performing 
         operations that eliminate adjacent equal elements by replacing them 
         with values within [1, w], ensuring no adjacent elements are equal.
Approach:
    - If no adjacent equal elements exist, return sum and 0 operations.
    - Otherwise:
        - Max sum is calculated as w * n - 1 (if possible).
        - Minimum operations are determined through analysis of segments 
          and specific pattern recognition.

Time Complexity: O(n)
Space Complexity: O(n)

The algorithm works by:
1. Identifying if any adjacent equal elements exist.
2. Computing the maximum achievable sum.
3. Evaluating the minimum number of operations required using
   dynamic considerations on segments of the array.
"""

import sys

input = sys.stdin.read
data = input().split()
index = 0


def fast_input():
    global index
    value = data[index]
    index += 1
    return value


def main():
    ta = int(fast_input())
    for _ in range(ta):
        n = int(fast_input())
        w = int(fast_input())
        a = [int(fast_input()) for _ in range(n)]
        can = 0
        for i in range(1, n):
            if a[i] == a[i - 1]:
                can = 1
        total_sum = sum(a)
        if can == 0:
            print(total_sum, 0)
            continue
        ans = w * n - 1
        if total_sum >= ans:
            print(total_sum, 0)
            continue
        if w == 2:
            k = 0
            l = 0
            for i in range(n):
                if a[i] == 1 and (i == 0 or a[i - 1] == 2):
                    k += 1
                if i > 0 and a[i] == 1 and a[i - 1] == 1:
                    l += 1
            print(2 * n - k, l)
            continue
        dist = 2 * n
        if (a[0] == a[1] and a[0] == w) or (a[n - 1] == a[n - 2] and a[n - 1] == w):
            dist = 0
        while a and a[-1] == w:
            a.pop()
        a.reverse()
        while a and a[-1] == w:
            a.pop()
        pen = 0
        n = len(a)
        cnt = 0
        for x in a:
            if x == w:
                cnt += 1
            else:
                if cnt == 1:
                    pen += 2
                else:
                    pen += (cnt + 1) // 2
                cnt = 0
        if cnt == 1:
            pen += 2
        else:
            pen += (cnt + 1) // 2
        ans3 = 10 * n + 1000000000
        if dist == 0:
            ans3 = n + 1 + pen
        i = 1
        c1 = 0
        while i < n and a[i] != a[i - 1]:
            if a[i - 1] == w:
                c1 += 1
            i += 1
        ans1 = ans3
        dec = 0
        if (
            i > 1
            and i + 1 < n
            and a[i] != w
            and a[i + 1] == w
            and a[i + 2] != w
            and (i == 1 or a[i - 2] != w)
        ):
            dec = 1
        if i < n:
            ans1 = n - 1 + pen - c1 * 2 + i - 1 - int(a[i] == w) - dec
        a.reverse()
        i = 1
        c1 = 0
        while i < n and a[i] != a[i - 1]:
            if a[i - 1] == w:
                c1 += 1
            i += 1
        ans2 = ans3
        dec = 0
        if (
            i > 1
            and i + 1 < n
            and a[i] != w
            and a[i + 1] == w
            and a[i + 2] != w
            and (i == 1 or a[i - 2] != w)
        ):
            dec = 1
        if i < n:
            ans2 = n - 1 + pen - c1 * 2 + i - 1 - int(a[i] == w) - dec
        print(ans, min(ans1, ans2, ans3))


if __name__ == "__main__":
    main()


# https://github.com/VaHiX/CodeForces/

File: CPP/ByRound/2053/2053_H_Delicate_Anti_monotonous_Operations.cpp
// Problem: CF 2053 H - Delicate Anti-monotonous Operations
// https://codeforces.com/contest/2053/problem/H

/*
 * Problem: H. Delicate Anti-monotonous Operations
 * Algorithms/Techniques: Simulation, Greedy, Two Pointers, Prefix Sum
 *
 * Time Complexity: O(n) per test case (amortized)
 * Space Complexity: O(1) additional space (excluding input array)
 *
 * Description:
 * Iris has an integer sequence of length n, each number in the range [1,w].
 * She can perform operations: select two adjacent equal elements a[i] = a[i+1],
 * and transform them to any values in [1,w], such that after operation a[i] != a[i+1].
 * Goal:
 *   Maximize the final sum of sequence.
 *   Minimize number of operations to reach this sum.
 *
 * Approach:
 * - If array already has no adjacent equal elements, answer is just sum and 0 operations.
 * - If any element in the array equals w, we can make all elements w (to maximize total).
 *   Then, we compute how many moves are needed.
 * - Otherwise, use greedy to try to avoid creating long consecutive sequences.
 * - For cases where we must break chains of repeated values:
 *   - Try different starting/ending points and simulate operation costs.
 *   - Reverse the subarray and see if it reduces operation count.
 */

#include <stdio.h>
#include <algorithm>
#include <iostream>
#include <numeric>

#define int long long
using namespace std;

inline int read() {
  int s = 0, w = 1;
  char ch = getchar();
  while (ch < '0' || ch > '9') {
    if (ch == '-')
      w = -1;
    ch = getchar();
  }
  while (ch <= '9' && ch >= '0')
    s = s * 10 + ch - '0', ch = getchar();
  return s * w;
}

int n, w, arr[200005];

// Calculates minimal operations to "break" a group of consecutive elements
int calculate_operations(int length) {
  if (length == 0)
    return 0;
  if (length == 1)
    return 2;
  return (length + 1) / 2;
}

// Finds minimum cost for subarray from left to right
int find_min_operations(int left, int right) {
  int consecutive_count = 0, operation_count = 0, min_operations = 1e18;
  for (int i = left; i < right; i++) {
    if (arr[i + 1] == arr[i]) {
      // Early termination condition: when current pair is equal
      min_operations =
          min(min_operations, operation_count + (right - i) + right - left - 1 -
                                  (consecutive_count == 1 && i + 2 <= right &&
                                   arr[i + 2] != w));
    }
    if (arr[i] == w) {
      operation_count -= calculate_operations(consecutive_count);
      consecutive_count++;
      operation_count += calculate_operations(consecutive_count);
    } else {
      consecutive_count = 0;
    }
  }
  return min_operations;
}

void solve(int test_case) {
  cin >> n >> w;
  for (int i = 1; i <= n; i++)
    cin >> arr[i];

  if (*min_element(arr + 1, arr + 1 + n) == w) {
    cout << n * w << " 0\n";
    return;
  }
  if (n == 1) {
    cout << arr[1] << " 0\n";
    return;
  }

  if (w == 2) {
    int max_sum = 2 * n, operations = 0;
    for (int i = 1; i <= n; i++)
      if (arr[i] == 1) {
        int left = i;
        while (left + 1 <= n && arr[left + 1] == arr[i])
          left++;
        max_sum--;
        operations += left - i;
        i = left;
      }
    cout << max_sum << " " << operations << "\n";
    return;
  }

  bool has_equal_adjacent = false;
  for (int i = 1; i < n; i++)
    if (arr[i] == arr[i + 1])
      has_equal_adjacent = true;
  if (!has_equal_adjacent) {
    cout << accumulate(arr + 1, arr + 1 + n, 0ll) << " 0\n";
    return;
  }
  cout << n * w - 1 << " ";
  if (accumulate(arr + 1, arr + 1 + n, 0ll) == n * w - 1) {
    cout << "0\n";
    return;
  }
  int left = (arr[1] == w ? 2 : 1), right = (arr[n] == w ? n - 1 : n);
  int min_operations = 1e18;
  if ((arr[n - 1] == w && arr[n] == w) || (arr[1] == w && arr[2] == w)) {
    while (arr[left] == w)
      left++;
    while (arr[right] == w)
      right--;
    int operations = right - left + 2;
    for (int i = left; i <= right; i++)
      if (arr[i] == w) {
        int x = i;
        while (x + 1 <= right && arr[x + 1] == w)
          x++;
        operations += calculate_operations(x - i + 1);
        i = x;
      }
    min_operations = operations;
  }
  min_operations = min(min_operations, find_min_operations(left, right));
  reverse(arr + left, arr + right + 1);
  min_operations = min(min_operations, find_min_operations(left, right));
  cout << min_operations << "\n";
}

signed main() {
  ios::sync_with_stdio(false);
  cin.tie(nullptr);
  int test_cases;
  cin >> test_cases;
  for (int i = 1; i <= test_cases; i++)
    solve(i);
  return 0;
}


// https://github.com/VaHiX/codeForces/